File: src/main/java/org/gephi/layout/plugin/forceAtlas2/ForceAtlas2.java
/*
 Copyright 2008-2011 Gephi
 Authors : Mathieu Jacomy <mathieu.jacomy@gmail.com>
 Website : http://www.gephi.org

 This file is part of Gephi.

 DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.

 Copyright 2011 Gephi Consortium. All rights reserved.

 The contents of this file are subject to the terms of either the GNU
 General Public License Version 3 only ("GPL") or the Common
 Development and Distribution License("CDDL") (collectively, the
 "License"). You may not use this file except in compliance with the
 License. You can obtain a copy of the License at
 http://gephi.org/about/legal/license-notice/
 or /cddl-1.0.txt and /gpl-3.0.txt. See the License for the
 specific language governing permissions and limitations under the
 License.  When distributing the software, include this License Header
 Notice in each file and include the License files at
 /cddl-1.0.txt and /gpl-3.0.txt. If applicable, add the following below the
 License Header, with the fields enclosed by brackets [] replaced by
 your own identifying information:
 "Portions Copyrighted [year] [name of copyright owner]"

 If you wish your version of this file to be governed by only the CDDL
 or only the GPL Version 3, indicate your decision by adding
 "[Contributor] elects to include this software in this distribution
 under the [CDDL or GPL Version 3] license." If you do not indicate a
 single choice of license, a recipient has the option to distribute
 your version of this file under either the CDDL, the GPL Version 3 or
 to extend the choice of license to its licensees as provided above.
 However, if you add GPL Version 3 code and therefore, elected the GPL
 Version 3 license, then the option applies only if the new code is
 made subject to such option by the copyright holder.

 Contributor(s):
 Alina Shlegel
 Timofey Pushkin

 Portions Copyrighted 2011 Gephi Consortium.


 Changes (made in 2021) compared to the original file:
 - Interfaces Node, Edge, Graph, and GraphModel are replaced with classes
 Node, Edge, and Graph (written on Kotlin) partially implementing their
 functionality.
 - Graph class mentioned above does not implement readLock / readUnlock
 methods, so their usage is removed.
 - Interfaces Layout, LayoutBuilder and class LayoutProperty are removed,
 so is the functionality related to them.
 - AbstractLayout interface usage is replaced with calculations from its
 method ensureSafeLayoutNodePositions.
 - Functionality related to the package org.openide.util is removed.
 - Dynamic edge weight support is removed.
 - Thread pool is modified to create daemon threads.
 */
package org.gephi.layout.plugin.forceAtlas2;

import org.gephi.graph.api.Edge;
import org.gephi.graph.api.Graph;
import org.gephi.graph.api.Node;
import org.gephi.layout.plugin.forceAtlas2.ForceFactory.AttractionForce;
import org.gephi.layout.plugin.forceAtlas2.ForceFactory.RepulsionForce;

import java.util.ArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * ForceAtlas 2 Layout, manages each step of the computations.
 *
 * @author Mathieu Jacomy
 */
public class ForceAtlas2 {

    private Graph graph;
    private double edgeWeightInfluence;
    private double jitterTolerance;
    private double scalingRatio;
    private double gravity;
    private double speed;
    private double speedEfficiency;
    private boolean outboundAttractionDistribution;
    private boolean adjustSizes;
    private boolean barnesHutOptimize;
    private double barnesHutTheta;
    private boolean linLogMode;
    private boolean strongGravityMode;
    private int threadCount;
    private int currentThreadCount;
    private Region rootRegion;
    double outboundAttCompensation = 1;
    private ExecutorService pool;

    public ForceAtlas2() {
        this.threadCount = Math.min(4, Math.max(1, Runtime.getRuntime().availableProcessors() - 1));
    }

    public void initAlgo() {
        speed = 1.;
        speedEfficiency = 1.;

        Node[] nodes = graph.getNodes().toArray(new Node[0]);

        // Initialise layout data
        for (Node n : nodes) {
            n.setX(((0.01 + Math.random()) * 1000.0) - 500.0);
            n.setY(((0.01 + Math.random()) * 1000.0) - 500.0);

            if (n.getLayoutData() == null || !(n.getLayoutData() instanceof ForceAtlas2LayoutData)) {
                ForceAtlas2LayoutData nLayout = new ForceAtlas2LayoutData();
                n.setLayoutData(nLayout);
            }
            ForceAtlas2LayoutData nLayout = n.getLayoutData();
            nLayout.mass = 1 + graph.getDegree(n);
            nLayout.old_dx = 0;
            nLayout.old_dy = 0;
            nLayout.dx = 0;
            nLayout.dy = 0;
        }

        pool = Executors.newFixedThreadPool(threadCount, (Runnable r) -> {
            Thread t = Executors.defaultThreadFactory().newThread(r);
            t.setDaemon(true);
            return t;
        });
        currentThreadCount = threadCount;
    }

    private double getEdgeWeight(Edge edge) {
        return edge.getWeight();
    }


    public void goAlgo() {
        // Initialize graph data
        Node[] nodes = graph.getNodes().toArray(new Node[0]);
        Edge[] edges = graph.getEdges().toArray(new Edge[0]);

        // Initialise layout data
        for (Node n : nodes) {
            if (n.getLayoutData() == null || !(n.getLayoutData() instanceof ForceAtlas2LayoutData)) {
                ForceAtlas2LayoutData nLayout = new ForceAtlas2LayoutData();
                n.setLayoutData(nLayout);
            }
            ForceAtlas2LayoutData nLayout = n.getLayoutData();
            nLayout.mass = 1 + graph.getDegree(n);
            nLayout.old_dx = nLayout.dx;
            nLayout.old_dy = nLayout.dy;
            nLayout.dx = 0;
            nLayout.dy = 0;
        }

        // If Barnes Hut active, initialize root region
        if (isBarnesHutOptimize()) {
            rootRegion = new Region(nodes);
            rootRegion.buildSubRegions();
        }

        // If outboundAttractionDistribution active, compensate.
        if (isOutboundAttractionDistribution()) {
            outboundAttCompensation = 0;
            for (Node n : nodes) {
                ForceAtlas2LayoutData nLayout = n.getLayoutData();
                outboundAttCompensation += nLayout.mass;
            }
            outboundAttCompensation /= nodes.length;
        }

        // Repulsion (and gravity)
        // NB: Muti-threaded
        RepulsionForce Repulsion = ForceFactory.builder.buildRepulsion(isAdjustSizes(), getScalingRatio());

        int taskCount = 8 * currentThreadCount;  // The threadPool Executor Service will manage the fetching of tasks and threads.
        // We make more tasks than threads because some tasks may need more time to compute.
        ArrayList<Future> threads = new ArrayList();
        for (int t = taskCount; t > 0; t--) {
            int from = (int) Math.floor(nodes.length * (t - 1) / taskCount);
            int to = (int) Math.floor(nodes.length * t / taskCount);
            Future future = pool.submit(new NodesThread(nodes, from, to, isBarnesHutOptimize(), getBarnesHutTheta(), getGravity(), (isStrongGravityMode()) ? (ForceFactory.builder.getStrongGravity(getScalingRatio())) : (Repulsion), getScalingRatio(), rootRegion, Repulsion));
            threads.add(future);
        }
        for (Future future : threads) {
            try {
                future.get();
            } catch (Exception e) {
                throw new RuntimeException("Unable to layout " + this.getClass().getSimpleName() + ".", e);
            }
        }

        // Attraction
        AttractionForce Attraction = ForceFactory.builder.buildAttraction(isLinLogMode(), isOutboundAttractionDistribution(), isAdjustSizes(), 1 * ((isOutboundAttractionDistribution()) ? (outboundAttCompensation) : (1)));
        if (getEdgeWeightInfluence() == 0) {
            for (Edge e : edges) {
                Attraction.apply(e.getSource(), e.getTarget(), 1);
            }
        } else if (getEdgeWeightInfluence() == 1) {
            for (Edge e : edges) {
                Attraction.apply(e.getSource(), e.getTarget(), getEdgeWeight(e));
            }
        } else {
            for (Edge e : edges) {
                Attraction.apply(e.getSource(), e.getTarget(), Math.pow(getEdgeWeight(e), getEdgeWeightInfluence()));
            }
        }

        // Auto adjust speed
        double totalSwinging = 0d;  // How much irregular movement
        double totalEffectiveTraction = 0d;  // Hom much useful movement
        for (Node n : nodes) {
            ForceAtlas2LayoutData nLayout = n.getLayoutData();
            if (!n.isFixed()) {
                double swinging = Math.sqrt(Math.pow(nLayout.old_dx - nLayout.dx, 2) + Math.pow(nLayout.old_dy - nLayout.dy, 2));
                totalSwinging += nLayout.mass * swinging;   // If the node has a burst change of direction, then it's not converging.
                totalEffectiveTraction += nLayout.mass * 0.5 * Math.sqrt(Math.pow(nLayout.old_dx + nLayout.dx, 2) + Math.pow(nLayout.old_dy + nLayout.dy, 2));
            }
        }
        // We want that swingingMovement < tolerance * convergenceMovement

        // Optimize jitter tolerance
        // The 'right' jitter tolerance for this network. Bigger networks need more tolerance. Denser networks need less tolerance. Totally empiric.
        double estimatedOptimalJitterTolerance = 0.05 * Math.sqrt(nodes.length);
        double minJT = Math.sqrt(estimatedOptimalJitterTolerance);
        double maxJT = 10;
        double jt = jitterTolerance * Math.max(minJT, Math.min(maxJT, estimatedOptimalJitterTolerance * totalEffectiveTraction / Math.pow(nodes.length, 2)));

        double minSpeedEfficiency = 0.05;

        // Protection against erratic behavior
        if (totalSwinging / totalEffectiveTraction > 2.0) {
            if (speedEfficiency > minSpeedEfficiency) {
                speedEfficiency *= 0.5;
            }
            jt = Math.max(jt, jitterTolerance);
        }

        double targetSpeed = jt * speedEfficiency * totalEffectiveTraction / totalSwinging;

        // Speed efficiency is how the speed really corresponds to the swinging vs. convergence tradeoff
        // We adjust it slowly and carefully
        if (totalSwinging > jt * totalEffectiveTraction) {
            if (speedEfficiency > minSpeedEfficiency) {
                speedEfficiency *= 0.7;
            }
        } else if (speed < 1000) {
            speedEfficiency *= 1.3;
        }

        // But the speed shoudn't rise too much too quickly, since it would make the convergence drop dramatically.
        double maxRise = 0.5;   // Max rise: 50%
        speed = speed + Math.min(targetSpeed - speed, maxRise * speed);

        // Apply forces
        if (isAdjustSizes()) {
            // If nodes overlap prevention is active, it's not possible to trust the swinging mesure.
            for (Node n : nodes) {
                ForceAtlas2LayoutData nLayout = n.getLayoutData();
                if (!n.isFixed()) {

                    // Adaptive auto-speed: the speed of each node is lowered
                    // when the node swings.
                    double swinging = nLayout.mass * Math.sqrt((nLayout.old_dx - nLayout.dx) * (nLayout.old_dx - nLayout.dx) + (nLayout.old_dy - nLayout.dy) * (nLayout.old_dy - nLayout.dy));
                    double factor = 0.1 * speed / (1f + Math.sqrt(speed * swinging));

                    double df = Math.sqrt(Math.pow(nLayout.dx, 2) + Math.pow(nLayout.dy, 2));
                    factor = Math.min(factor * df, 10.) / df;

                    double x = n.x() + nLayout.dx * factor;
                    double y = n.y() + nLayout.dy * factor;

                    n.setX((float) x);
                    n.setY((float) y);
                }
            }
        } else {
            for (Node n : nodes) {
                ForceAtlas2LayoutData nLayout = n.getLayoutData();
                if (!n.isFixed()) {

                    // Adaptive auto-speed: the speed of each node is lowered
                    // when the node swings.
                    double swinging = nLayout.mass * Math.sqrt((nLayout.old_dx - nLayout.dx) * (nLayout.old_dx - nLayout.dx) + (nLayout.old_dy - nLayout.dy) * (nLayout.old_dy - nLayout.dy));
                    //double factor = speed / (1f + Math.sqrt(speed * swinging));
                    double factor = speed / (1f + Math.sqrt(speed * swinging));

                    double x = n.x() + nLayout.dx * factor;
                    double y = n.y() + nLayout.dy * factor;

                    n.setX((float) x);
                    n.setY((float) y);
                }
            }
        }
    }

    public boolean canAlgo() {
        return graph != null;
    }

    public void endAlgo() {
        for (Node n : graph.getNodes()) {
            n.setLayoutData(null);
        }
        pool.shutdown();
    }

    public void resetPropertiesValues() {
        int nodesCount = 0;

        if (graph != null) {
            nodesCount = graph.getNodeCount();
        }

        // Tuning
        if (nodesCount >= 100) {
            setScalingRatio(2.0);
        } else {
            setScalingRatio(10.0);
        }
        setStrongGravityMode(false);
        setGravity(1.);

        // Behavior
        setOutboundAttractionDistribution(false);
        setLinLogMode(false);
        setAdjustSizes(false);
        setEdgeWeightInfluence(1.);

        // Performance
        setJitterTolerance(1d);
        if (nodesCount >= 1000) {
            setBarnesHutOptimize(true);
        } else {
            setBarnesHutOptimize(false);
        }
        setBarnesHutTheta(1.2);
        setThreadsCount(Math.max(1, Runtime.getRuntime().availableProcessors() - 1));
    }

    public void setGraph(Graph graph) {
        this.graph = graph;
        // Trick: reset here to take the profile of the graph in account for default values
        resetPropertiesValues();
    }

    public Double getBarnesHutTheta() {
        return barnesHutTheta;
    }

    public void setBarnesHutTheta(Double barnesHutTheta) {
        this.barnesHutTheta = barnesHutTheta;
    }

    public Double getEdgeWeightInfluence() {
        return edgeWeightInfluence;
    }

    public void setEdgeWeightInfluence(Double edgeWeightInfluence) {
        this.edgeWeightInfluence = edgeWeightInfluence;
    }

    public Double getJitterTolerance() {
        return jitterTolerance;
    }

    public void setJitterTolerance(Double jitterTolerance) {
        this.jitterTolerance = jitterTolerance;
    }

    public Boolean isLinLogMode() {
        return linLogMode;
    }

    public void setLinLogMode(Boolean linLogMode) {
        this.linLogMode = linLogMode;
    }

    public Double getScalingRatio() {
        return scalingRatio;
    }

    public void setScalingRatio(Double scalingRatio) {
        this.scalingRatio = scalingRatio;
    }

    public Boolean isStrongGravityMode() {
        return strongGravityMode;
    }

    public void setStrongGravityMode(Boolean strongGravityMode) {
        this.strongGravityMode = strongGravityMode;
    }

    public Double getGravity() {
        return gravity;
    }

    public void setGravity(Double gravity) {
        this.gravity = gravity;
    }

    public Integer getThreadsCount() {
        return threadCount;
    }

    public void setThreadsCount(Integer threadCount) {
        this.threadCount = Math.max(1, threadCount);
    }

    public Boolean isOutboundAttractionDistribution() {
        return outboundAttractionDistribution;
    }

    public void setOutboundAttractionDistribution(Boolean outboundAttractionDistribution) {
        this.outboundAttractionDistribution = outboundAttractionDistribution;
    }

    public Boolean isAdjustSizes() {
        return adjustSizes;
    }

    public void setAdjustSizes(Boolean adjustSizes) {
        this.adjustSizes = adjustSizes;
    }

    public Boolean isBarnesHutOptimize() {
        return barnesHutOptimize;
    }

    public void setBarnesHutOptimize(Boolean barnesHutOptimize) {
        this.barnesHutOptimize = barnesHutOptimize;
    }
}

File: src/main/kotlin/org/gephi/layout/plugin/forceAtlas2/ForceAtlas2.kt
package org.gephi.layout.plugin.forceAtlas2

import ru.spbu.netter.model.Network
import java.util.concurrent.Executors
import java.util.concurrent.Future
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sqrt


class ForceAtlas2(
    network: Network,
    private val outboundAttractionDistribution: Boolean = true,
    private val adjustSizes: Boolean = true,
    private val barnesHutOptimize: Boolean = true,
    private val linLogMode: Boolean = false,
    private val strongGravityMode: Boolean = false,
    private val edgeWeightInfluence: Double = 1.0,
    private val jitterTolerance: Double = 1.0,
    private val scalingRatio: Double = 2.0,
    private val gravity: Double = 1.0,
    private val barnesHutTheta: Double = 1.0,
) {
    private val faNodes = mutableListOf<FaNode>()
    private val faLinks = mutableListOf<FaLink>()
    private val degrees = Array(network.nodes.size) { 0 }

    private var speed = 1.0
    private var speedEfficiency = 1.0
    private var outboundAttCompensation = 1.0
    private lateinit var rootRegion: Region

    private val threadCount = 4.coerceAtMost(1.coerceAtLeast(Runtime.getRuntime().availableProcessors() - 1))
    private val currentThreadCount = threadCount
    private val pool = Executors.newFixedThreadPool(threadCount)

    init {
        for (link in network.links) {
            faLinks += FaLink(link)

            degrees[link.n1.id]++
            if (link.n1.id != link.n2.id) degrees[link.n2.id]++
        }

        for (node in network.nodes.values) faNodes += FaNode(node).apply { mass = 1.0 + degrees[node.id] }
    }

    fun start() {
        for (faNode in faNodes) faNode.apply {
            mass = 1.0 + degrees[node.id]
            oldDx = dx
            oldDy = dy
            dx = 0.0
            dy = 0.0
        }

        if (barnesHutOptimize) rootRegion = Region(faNodes).apply { buildSubRegions() }

        if (outboundAttractionDistribution) {
            outboundAttCompensation = 0.0
            for (nodeLayout in faNodes) outboundAttCompensation += nodeLayout.mass
            outboundAttCompensation /= faNodes.size
        }

        val repulsionForce = ForceFactory().buildRepulsion(adjustSizes, scalingRatio)

        val taskCount = 8 * currentThreadCount
        val threads = mutableListOf<Future<*>>()
        for (taskNum in taskCount downTo 1) threads.add(
            pool.submit(
                NodesThread(
                    faNodes,
                    floor((faNodes.size * (taskNum - 1) / taskCount).toDouble()).toInt(),
                    floor((faNodes.size * taskNum / taskCount).toDouble()).toInt(),
                    barnesHutOptimize,
                    barnesHutTheta,
                    gravity,
                    if (strongGravityMode) ForceFactory().getStrongGravity(scalingRatio) else repulsionForce,
                    scalingRatio,
                    rootRegion,
                    repulsionForce,
                )
            )
        )
        for (future in threads) {
            try {
                future.get()
            } catch (ex: Exception) {
                throw RuntimeException("Unable to layout " + this.javaClass.simpleName + ".", ex)
            }
        }

        val attractionForce = ForceFactory().buildAttraction(
            linLogMode, outboundAttractionDistribution, adjustSizes,
            1.0 * if (outboundAttractionDistribution) outboundAttCompensation else 1.0,
        )
        when (edgeWeightInfluence) {
            0.0 -> for (faLink in faLinks) attractionForce.apply(
                faNodes[faLink.link.n1.id],
                faNodes[faLink.link.n2.id],
                1.0
            )
            1.0 -> for (faLink in faLinks) attractionForce.apply(
                faNodes[faLink.link.n1.id],
                faNodes[faLink.link.n2.id],
                faLink.weight
            )
            else -> for (faLink in faLinks) attractionForce.apply(
                faNodes[faLink.link.n1.id],
                faNodes[faLink.link.n2.id],
                faLink.weight.pow(edgeWeightInfluence)
            )
        }

        var totalSwinging = 0.0
        var totalEffectiveTraction = 0.0
        for (node in faNodes) {
            if (node.isFixed) continue
            totalSwinging += node.mass * sqrt((node.oldDx - node.dx).pow(2.0) + (node.oldDy - node.dy).pow(2.0))
            totalEffectiveTraction +=
                node.mass * 0.5 * sqrt((node.oldDx + node.dx).pow(2.0) + (node.oldDy + node.dy).pow(2.0))
        }

        val estimatedOptimalJitterTolerance = 0.05 * sqrt(faNodes.size.toDouble())
        val minJT = sqrt(estimatedOptimalJitterTolerance)
        val maxJT = 10.0
        var jt = jitterTolerance * minJT.coerceAtLeast(
            maxJT.coerceAtMost(
                estimatedOptimalJitterTolerance * totalEffectiveTraction / faNodes.size.toDouble().pow(2.0)
            )
        )

        val minSpeedEfficiency = 0.05

        if (totalSwinging / totalEffectiveTraction > 2.0) {
            if (speedEfficiency > minSpeedEfficiency) speedEfficiency *= 0.5
            jt = jt.coerceAtLeast(jitterTolerance)
        }

        val targetSpeed = jt * speedEfficiency * totalEffectiveTraction / totalSwinging

        if (totalSwinging > jt * totalEffectiveTraction) {
            if (speedEfficiency > minSpeedEfficiency) speedEfficiency *= 0.7
        } else if (speed < 1000) speedEfficiency *= 1.3

        val maxRise = 0.5
        speed += (targetSpeed - speed).coerceAtMost(maxRise * speed)

        if (adjustSizes) {
            for (node in faNodes) {
                if (node.isFixed) continue

                val swinging =
                    node.mass * sqrt((node.oldDx - node.dx) * (node.oldDx - node.dx) + (node.oldDy - node.dy) * (node.oldDy - node.dy))
                val df = sqrt(node.dx.pow(2.0) + node.dy.pow(2.0))
                val factor = (0.1 * speed / (1f + sqrt(speed * swinging)) * df).coerceAtMost(10.0) / df

                node.x = node.x + node.dx * factor
                node.y = node.y + node.dy * factor
            }
        } else {
            for (node in faNodes) {
                if (node.isFixed) continue

                val swinging: Double =
                    node.mass * sqrt((node.oldDx - node.dx) * (node.oldDx - node.dx) + (node.oldDy - node.dy) * (node.oldDy - node.dy))
                val factor = speed / (1f + sqrt(speed * swinging))

                node.x = node.x + node.dx * factor
                node.y = node.y + node.dy * factor
            }
        }
    }

    fun getResults(): List<FaNode> {
        pool.shutdown()
        return faNodes
    }
}
